app/main.py

```diff
@@ -1,214 +1,47 @@
-        },
-        "potion": None,
-    },
-    "mordred": {
-        "name": "Mordred",
-        "power": 30,
-        "hp": 90,
-        "armour": [
-            {
-                "part": "breastplate",
-                "protection": 15,
-            },
-            {
-                "part": "boots",
-                "protection": 10,
-            }
-        ],
-        "weapon": {
-            "name": "Poisoned Sword",
-            "power": 60,
-        },
-        "potion": {
-            "name": "Berserk",
-            "effect": {
-                "power": +15,
-                "hp": -5,
-                "protection": +10,
-            }
-        }
-    },
-    "red_knight": {
-        "name": "Red Knight",
-        "power": 40,
-        "hp": 70,
-        "armour": [
-            {
-                "part": "breastplate",
-                "protection": 25,
-            }
-        ],
-        "weapon": {
-            "name": "Sword",
-            "power": 45
-        },
-        "potion": {
-            "name": "Blessing",
-            "effect": {
-                "hp": +10,
-                "power": +5,
-            }
-        }
+from app.database.participants import KNIGHTS
+from app.knights.knight import Knight
+from app.knights.armour import Armour
+from app.knights.weapon import Weapon
+from app.knights.potion import Potion
+from app.arena.fight import Fight
+
+
+def battle(knights_config: dict) -> dict:
+    participants = {
+        name: Knight(knight["name"], knight["power"], knight["hp"])
+        for name, knight in knights_config.items()
     }
-}
 
+    for name, knight in participants.items():
 
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
+        if knights_config.get(name, {}).get("armour", []) is not None:
+            knight_armour = [
+                Armour(armour["part"], armour["protection"])
+                for armour in knights_config.get(name, {}).get("armour", [])
+            ]
 
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
+            for armour in knight_armour:
+                knight.apply_armour(armour)
 
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
+        weapon = Weapon(
+            knights_config.get(name, {}).get("weapon", {})["name"],
+            knights_config.get(name, {}).get("weapon", {})["power"]
+        )
 
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
+        knight.apply_weapon(weapon)
 
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
+        if knights_config.get(name, {}).get("potion", {}) is not None:
+            potion = Potion(
+                knights_config.get(name, {}).get("potion", {})["name"],
+                knights_config.get(name, {}).get("potion", {})["effect"]
```

Comment:
you can use a variable instead of code duplication.
Also you take "name" and "effect" without get, what if it would be not at your dict?